geode-core/src/main/java/org/apache/geode/internal/cache/InitialImageOperation.java

```diff
@@ -1878,6 +1881,16 @@ public boolean executeWith(Object entList, int b) {
           sendFailureMessage(dm, rex);
         } // !success
 
+        if (lostMemberID != null && targetRegion != null) {
+          if (lostMemberVersionID == null) {
+            lostMemberVersionID = lostMemberID;
+          }
+          // check to see if the region in this cache needs to synchronize with others
+          // it is possible that the cache is recover/restart of a member and not
+          // scheduled to synchronize with others
+          synchronizeIfNotScheduled(targetRegion, lostMemberID, lostMemberVersionID);
```

Comment:
I'm a little confused about what's going on.  Synchronization for a lost member can happen multiple times over the lifespan of a server.  Is the "scheduled"ness of synchronization reset anywhere?

Reply:
See the command above, each time a disk store id is lost, the restarted one will have an initialized false set to the flag.
If the member is not lost, the timed task of region sync will be triggered by the memberDeparted event, which ignore the flag setting.

Reply:
I had the same question before. But I discussed with Eric and i know his design. I put his design somewhere else. Now I will cut and paste here:

There’re 2 tricks here:

How to determine the GII requester?
when the current GII requester for lostMember restarted before the syncWith(lostMember), it will receive the syncWith(lostMember) from other live members. When processing this syncWith(lostMember) request, it’s the only member who did not schedule the syncWith operation. So it knows that it needs to send a special syncWith operation for the lost member(could be itself).

If a new member joined at this time, its holder for the lost member did not set syncWith, it will also call a special syncWith operation for the lost member.  

If the GII requester is restarted after syncWith(lostMember) was sent, it will not received the syncWith(lostMember), but its GII should have correct result. No need to make up.

 

 

2. Introduced a boolean regionSyncScheduledOrDone in RVHolder. If the syncWith is scheduled by membership listener, set it to true. Not to do additional syncWith for this holder.

If a member is restarted or newly joined, its regionSyncScheduledOrDone is false. We use this boolean to determine who should send additional syncWith operation.